apiserver/application/application.go

```diff
@@ -339,6 +360,100 @@ func (api *API) Update(args params.ApplicationUpdate) error {
+		return err
+	}
+	if !args.Force {
+		if err := api.check.ChangeAllowed(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	app, err := api.backend.Application(args.ApplicationName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if args.Series == "" {
+		return params.Error{
+			Message: "series missing from args",
+			Code:    params.CodeBadRequest,
+		}
+	}
+	if err = verifySupportedSeries(app, args.Series, args.Force); err != nil {
+		if params.IsCodeIncompatibleSeries(err) {
+			err = errors.Annotatef(err, "series %s is not supported by the application in use. Use --force to override.", args.Series)
+		}
+		return err
+	}
+
+	if !app.IsPrincipal() {
+		return errors.Errorf("%s is a subordinate application, series update is not supported.", args.ApplicationName)
+	}
+
+	units, err := app.AllUnits()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// All units will have the same subordinates, therefore we only
+	// need to check one.
+	if len(units) == 0 {
+		return errors.Errorf("no units of this application have been deployed, nothing to update.")
+	}
+	unit := units[0]
+	// Find our subordinate applications, to verify their series as well,
+	for _, subUnitName := range unit.SubordinateNames() {
+		subordinateApplicationName := strings.Split(subUnitName, "/")[0]
+		subordinateApplication, err := api.backend.Application(subordinateApplicationName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if err = verifySupportedSeries(subordinateApplication, args.Series, args.Force); err != nil {
+			if params.IsCodeIncompatibleSeries(err) {
+				err = errors.Annotatef(err, "series %s is not supported by the subordinates of the application in use. Use --force to override.", args.Series)
+			}
+			return err
+		}
+	}
+
+	// Update the series of the subordinate applications as well.  When
+	// only the principle application is updated and add-unit is called,
+	// the subordinate applications are still installed.  This can
+	// make for inconsitent data.
+	//
+	// Note: It is possible for a subordinate charm to have more than one
+	// principal.  However it doesn't appear that the Series data in the doc
+	// is used.  Updating to match the principal used here at least.
+	for _, subUnitName := range unit.SubordinateNames() {
+		subordinateApplicationName := strings.Split(subUnitName, "/")[0]
+		subordinateApplication, err := api.backend.Application(subordinateApplicationName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = subordinateApplication.UpdateApplicationSeries(args.Series)
+		if err != nil {
+			return err
+		}
+	}
+
+	return app.UpdateApplicationSeries(args.Series)
+}
+
+func verifySupportedSeries(app Application, series string, force bool) error {
+	ch, _, err := app.Charm()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	_, seriesSupportedErr := charm.SeriesForCharm(series, ch.Meta().Series)
+	if seriesSupportedErr != nil && !force {
+		return errors.New(params.CodeIncompatibleSeries)
```

Comment:
This isn't right. It's somethings like
return &params.Error{
    Code: params.CodeIncompatibleSeries,
    Message: "blah",
    etc
}

This implies there's something wrong with the api error handling as well as that should have been broken by this.